### PR TITLE
Add RecordException overload accepting additional tags

### LINKED ISSUE #3369: RecordException needs to allow for adding additional event attributes
Originally raised here https://github.com/open-telemetry/opentelemetry-dotnet/issues/2439#issuecomment-1001747365.

[The specification](https://github.com/open-telemetry/opentelemetry-specification/blob/5fb819c72e34cb21fe1ce5acba1dc93c2216ecb1/specification/trace/api.md#record-exception) states:

> If RecordException is provided, the method MUST accept an optional parameter to provide any additional event attributes (this SHOULD be done in the same way as for the AddEvent method). If attributes with the same name would be generated by the method already, the additional attributes take precedence.

### PR DESCRIPTION
**Pull request by alanwest**
Fixes #3369.

Adds `RecordException` overload accepting additional tags to add to the exception event.

```csharp
public static void RecordException(this Activity activity, Exception ex, TagList tags)
```

### CODE REVIEW

**Comment by CodeBlanch** [comment]
@alanwest Could we use [TagList](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.taglist?view=net-6.0)?

Reasons...

* Allocation free up to 8 tags.
* Allocation free foreach.
* Slightly nicer syntax (debatable 🤣) you can new it up using collection initializer like...

   ```csharp
   activity.RecordException(ex, new() { new("myTag", "myValue") })
   ```

If you do decide to go with `TagList` it could be large so we should probably use `in` modifier:

   ```csharp
    public static void RecordException(this Activity activity, Exception ex, in TagList tags)
   ```

**Comment by cijothomas** [comment]
> Also, one might argue that RecordException should be offered by .NET itself since it is part of the spec and not as an extension method offered by us.

The spec is not marked stable for the Exceptions part.!

**Comment by cijothomas** [comment]
> I didn't reach for TagList is that it is not used by any of the existing Activity APIs (might fact check me on this...)

IIRC, TagList was added while adding Metrics. It was intentionally not placed in .Metrics namespace, to re-use it in the future for Tracing as well. Its something we need to ask to be supported by `StartActivity` overloads. (We have a good use-case now - spec is recommending to add all available attributes at span creation. One good way to do it alloc-free would be to add more overloads to StartActivity, accepting `TagList`...

**Comment by CodeBlanch** [comment]
> I agree with the benefits, but I guess the reason I didn't reach for TagList is that it is not used by any of the existing Activity APIs

Ya like @cijothomas said I think that is just because it came later 😄 

> Also, one might argue that RecordException should be offered by .NET itself since it is part of the spec and not as an extension method offered by us.

I think someday it will get there 😄 https://github.com/dotnet/runtime/issues/53641

> TagList implements IEnumerable<KeyValuePair<string, object>>, so it could still be used - albeit not with the sugary syntax. Could RecordException do a tags is TagList enumerateMeForFree to get the perf benefit? I guess there's probably a boxing in passing it to RecordException, but then would there still be a benefit when enumerating?

You mean have the method accept the interface and let user pass in `TagList` if they so-desire? Yes, possible, but allocates + copies so probably worse for perf!

If you made it `RecordException<T>(Exception ex, T tags) where T : IEnumerable<KeyValuePair<string, object>>` you could probably avoid the boxing and then use `is TagList` in the implementation to avoid the enumerator.

**Comment by alanwest** [comment]
> The spec is not marked stable for the Exceptions part.!

and

> I think someday it will get there 😄 https://github.com/dotnet/runtime/issues/53641

Understood - though wasn't aware of the already open issue. The thing I was unsure of was whether once it was part of the Activity API what type would be used. Sounds like it's expected that `TagList` will be used. If so, then yes makes complete sense to use TagList.

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs [hunk 1]
```diff
@@ -92,6 +104,11 @@ public static void RecordException(this Activity activity, Exception ex)
                 tagsCollection.Add(SemanticConventions.AttributeExceptionMessage, ex.Message);
             }
 
+            foreach (var tag in tags)
```
Comment (utpilla) [inline src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:107]: Add a `null` check for `activity` in the beginning to `return` if the `activity` is `null` and avoid creating `tagsCollection` and copying from `tagList`.

**Review by utpilla — APPROVED** [review]
Left a suggestion. Looks good overall.

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs (hunk 1 above)
Comment (alanwest) [reply src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:107]: Added null check.

This brought my attention to a somewhat interesting question: what if `ex` is null, but `tags` is not?

One might argue that the only reason someone would call `RecordException` is if they had a non-null exception and that if they just wanted to add an event then they'd call `Activity.AddEvent`.

But what do folks think? Should we add an event anyways even though no [exception related attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md#attributes) may end up on the event?

Comment (utpilla) [reply src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:107]: > Should we add an event anyways even though no [exception related attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md#attributes) may end up on the event?

It doesn't sound right to me. I assume most people would use this in the `catch` block or some callback for when an exception occurs. So, it should not be a common scenario. I would say, as you mentioned, they should just use `Activity.AddEvent` if it has nothing to do with an actual exception.

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs [hunk 2]
```diff
@@ -92,6 +104,11 @@ public static void RecordException(this Activity activity, Exception ex)
...
+            {
+                tagsCollection[tag.Key] = tag.Value;
+            }
+
             activity?.AddEvent(new ActivityEvent(SemanticConventions.AttributeExceptionEventName, default, tagsCollection));
```
Comment (utpilla) [inline src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:112]: nit: we don't need the `?` operator

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs [hunk 3]
```diff
@@ -69,14 +69,26 @@ public static Status GetStatus(this Activity activity)
...
+        }
+
+        /// <summary>
+        /// Adds an activity event containing information from the specified exception and additional tags.
+        /// </summary>
+        /// <param name="activity">Activity instance.</param>
+        /// <param name="ex">Exception to be recorded.</param>
+        /// <param name="tags">Additional tags to record on the event.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void RecordException(this Activity activity, Exception ex, in TagList tags)
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:89]: unless Activity.AddEvent/team also accepts TagList, the benefits of using TagList in this API is nil, as we anyway have to allocate the ActivityTagsCollection.

Hopefully if/when the spec is stable, we can bring this up.

**Review by cijothomas — APPROVED** [review]
Please update the PR description with the final shape of the API.

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs (hunk 3 above)
Comment (alanwest) [reply src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:89]: I think the idea is that the consumer can still avoid some allocations when using `TagList`, but yea the benefit would be greater if `AddEvent` accepted TagList as well.

**Pull request merged**